### PR TITLE
Proposal: Start server simplification

### PR DESCRIPTION
**Pull request by andrewemeryanz**
Reduce server start to a single function:

```go
customConfig := CustomConfig{}
return StartServer(file, &customConfig, func(cfg *config.DefaultConfig) (*Server, error) {
    serviceInterface := ServiceInterface{}
    return NewServer(ctx, cfg).
        WithRest(&serviceInterface, &CustomRestCallback{}).
        WithGrpc(&serviceInterface, &CustomGrpcCallback{}).
        WithLogHook(&CustomLogHook{}), nil
    })
```

The proposal is to generate a custom `Server` struct and builder to simplify server initialisation:

```go
type Server struct {
    params           *core.ServerParams
    serviceInterface *ServiceInterface
    restGenCallback  core.RestGenCallback
    grpcGenCallback  core.GrpcGenCallback
}

func NewServer(ctx context.Context, config *config.DefaultConfig) *Server { ... }
func (s *Server) WithRest(serviceInterface *ServiceInterface, callback core.RestGenCallback) *Server { ... }
func (s *Server) WithGrpc(serviceInterface *ServiceInterface, callback core.GrpcGenCallback) *Server { ... }
```

This builder can be used directly if consumers pass in the server configuration directly:

```go
customConfig := CustomConfig{}
cfg := NewDefaultConfig()
err := config.LoadConfig(file, &cfg, &customConfig)
    if err != nil {
        return err
    }
serviceInterface := ServiceInterface{}
NewServer(ctx, cfg).
    WithRest(&serviceInterface, &CustomRestCallback{}).
    WithGrpc(&serviceInterface, &CustomGrpcCallback{}).
    WithLogHook(&CustomLogHook{}).
    Start()
```

Or the server configuration can be hidden entirely through the use of the generated `StartServer` method:

```go
customConfig := CustomConfig{}
return StartServer(file, &customConfig, func(cfg *config.DefaultConfig) (*Server, error) {
    serviceInterface := ServiceInterface{}
    return NewServer(ctx, cfg).
        WithRest(&serviceInterface, &CustomRestCallback{}).
        WithGrpc(&serviceInterface, &CustomGrpcCallback{}).
        WithLogHook(&CustomLogHook{}), nil
    })
```

### CODE REVIEW

**Comment by anzdaddy** [comment]
I would prefer:

1. A more strongly type model, whereby `Server` knows about all endpoints.
2. Pass a context through the callback for non-essential tweaks such as log-hooks.

Something like:

```go
return StartServer(&customConfig, func(ctx context.Context, cfg *config.DefaultConfig) (ctx context.Context, *Server, error) {
    ctx = log.WithLogger(ctx, NewMyLogger())
    return ctx, &Server{
        foo: func(...) ... {
            ...
        },
        bar: func(...) ... {
            ...
        },
    }, nil
})
```

File: codegen/tests/simple/server.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+type Server struct {
+	params           *core.ServerParams
+	serviceInterface *ServiceInterface
+	restGenCallback  core.RestGenCallback
+	grpcGenCallback  core.GrpcGenCallback
+}
+
+func StartServer(file string, customConfig interface{}, build func(*config.DefaultConfig) (*Server, error)) error {
+	cfg := NewDefaultConfig()
+	err := config.LoadConfig(file, &cfg, &customConfig)
```
Comment (ChloePlanet) [inline codegen/tests/simple/server.go:24]:
```suggestion
	err := config.LoadConfig(file, &cfg, customConfig)
```
IMO the passed in customConfig supposed to be the address of the custom config variable.

Comment (andrewemeryanz) [reply codegen/tests/simple/server.go:24]: Thanks. I suspect you're right. The proposal is just structural. I haven't actually executed it :)

**Comment by andrewemeryanz** [comment]
> 1. A more strongly type model, whereby `Server` knows about all endpoints.

Endpoints are currently encapsulated within the generated `ServiceInterface` struct:
```go
si := ServiceInterface{
    foo: ...,
    bar: ...,
}
```
It's possible but it would require some refactoring. What's the benefit of exploding these here?

```go
return ctx, &Server{
    foo: func(...) ... {
    },
    bar: func(...) ... {
    },
}, nil
```

**Comment by anzdaddy** [comment]
Maybe I've misunderstood the intent. Is `WithRest` a generic method for REST endpoints or an example of a specific endpoint? If it's the latter, then I'm more comfortable with it. However, passing `serviceInterface` in each call to set the same underlying field is a bit smelly.

**Comment by andrewemeryanz** [comment]
@anzdaddy `WithRest` is poorly named on my part. It's saying: expose this server as a REST service.

**Comment by anzdaddy** [comment]
After a conversation with @andrewemeryanz to clarify things, here's my understanding.

1. `WithRest` is a generic method to wire up the code-generated interface, likewise for `WithGrpc`. This is fine, because the interface itself is strongly typed.
1. The second parameter to these methods, `Custom*Callback`, is the bit that threw me and in the discussion @andrewemeryanz observed that the name is a little confusing, since it's really a customisation mechanism that the framework calls in order to tailor certain behaviours.
1. We agreed that that second parameter could be replaced with something like `options ...core.ServerOption` following the popular configuration idiom exemplified by [grpc.Invoke](https://godoc.org/google.golang.org/grpc#Invoke)(…, opts ...[CallOption](https://godoc.org/google.golang.org/grpc#CallOption)).
1. We also concluded that it would be useful for `NewServer` to accept the same variadic parameter so customisations that apply to both protocols can be specified once.
1. The `WithLogHook` method can be recast as a `core.ServerOption`.

A couple of points occurred to me after the discussion:

1. Serving both REST and gRPC might be a very common choice, so perhaps `WithRestAndGrpc` is worth introducing.
1. `ctx` seems to come from nowhere. It should be passed to the callback.
1. I would prefer `config.DefaultConfig` to be invisible and therefore not passed through the callback.
1. The custom config type can be inferred from the callback function signature via reflection instead of being passed in as a pointer parameter.

I'm hoping we can arrive at something closer to this:

```go
return StartServer(configFile, func(ctx context.Context, cfg *CustomConfig) (*Server, error) {
    ctx = log.WithLogger(NewMyLogger()).Onto(ctx)
    return NewServer(ctx, core.BasePath("/mybase")).WithRestAndGrpc(&ServiceInterface{})
})
```

**Comment by andrewemeryanz** [comment]
Unrequired.

**Pull request closed without merging**